Fix chi2 and errors in fit function

bpalmeiro opened the pull request:
- Chi2
Implemented chi2 computation, based on scipy.stat.chisquare is only valid when data smearing is Poisson distributed. So, a general chi2 function is added, with formula extracted from [1].

- Errors
Fit errors, given by scipy.optimize.curve_fit were not correct as flag absolute_sigma was set to false. When true, "[data] sigma describes one standard deviation errors of the input data points", so covariance matrix estimation is based on these values. Otherwise, data sigma is only taken as "relative weights of the data points" and covariance matrix is not affected by the overall magnitude of the values in sigma, just by relative magnitudes.

Both improvements are shown in https://github.com/bpalmeiro/ICARO/blob/Chi2_demostration/icaro/Chi2_demonstration/Chi2_demonstration.ipynb

[1]  Numerical Recipes. The Art of Scientific Computing, 3rd Edition, 2007, ISBN 0-521-88068-8. Pages 778-780

gonzaponte commented on invisible_cities/core/fit_functions.py:
```diff
@@ -27,6 +27,35 @@ def get_errors(cov):
+        Fit funtion values corresponding to ydata array
+    sigma : np.ndarray
+        Data errors array. If no sigma is given, it takes the poisson case.
+
```
ndf is not documented

gonzaponte replied:
I don't want to be pedantic, but can we replace `len(data)` by `number of data points`?

gonzaponte commented on invisible_cities/core/fit_functions.py:
```diff
@@ -96,16 +125,26 @@ def fit(func, x, y, seed=(), fit_range=None, **kwargs):
     fitf = lambda x: func(x, *vals)
     fitx = fitf(x)
-    chi2, pval = scipy.stats.chisquare(y[fitx != 0], fitx[fitx != 0], len(vals))
+
```
use a variable to hold the result of `fitx != 0`. It is used in 4 different places.

gonzaponte commented on invisible_cities/core/fit_functions.py:
```diff
@@ -96,16 +125,26 @@ def fit(func, x, y, seed=(), fit_range=None, **kwargs):
-    chi2, pval = scipy.stats.chisquare(y[fitx != 0], fitx[fitx != 0], len(vals))
+
+    if not "sigma" in kwargs:
+        sigma_r   = None
```
You can use `sigma_r = kwargs.get("sigma")` which returns `None` if the key is not found.  Or `sigma_r = kwargs.get("sigma", None)`, which is more explicit.

gonzaponte reviewed (changes requested):
Great work! Very well documented and demonstrated.

gonzaponte commented on invisible_cities/reco/corrections_test.py:
```diff
@@ -1,13 +1,14 @@
+from ..reco.corrections    import LifetimeCorrection
+from ..reco.corrections    import LifetimeRCorrection
+from ..reco.corrections    import LifetimeXYCorrection
+from ..icaro.hst_functions import poisson_sigma
```
Please move `poisson_sigma` to `core_functions.py`. The `icaro` package will (probably) be removed in the near future.

gonzaponte commented on invisible_cities/core/fit_functions.py:
```diff
@@ -137,8 +137,8 @@ def fit(func, x, y, seed=(), fit_range=None, **kwargs):
-    sigma_r = kwargs.get("sigma", None)
-    if sigma_r is not None:
+    sigma_r = kwargs.get("sigma", 1)
+    if np.size(sigma_r) is not 1:
```
Mmm, I don't like this `size` check... How about setting the default to `np.ones` of the corresponding size and applying the `nonzero` selection over it?

gonzaponte commented on invisible_cities/reco/corrections_test.py:
```diff
@@ -360,9 +361,10 @@ def test_corrections_1d(gauss_data_1d):
-    x    = x[:-1] + np.diff(x) * 0.5
-    f    = fitf.fit(fitf.gauss, x, y, (1e5, mean, std))
+    y, x  = np.histogram(Eevt, np.linspace(mean - 3 * std, mean + 3 * std, 100))
+    x     = x[:-1] + np.diff(x) * 0.5
```
This is the same as the function `shift_to_bin_centers` in the `icaro` package. Can we move it also to `core` and use it here?

gonzaponte commented on invisible_cities/reco/corrections_test.py:
```diff
@@ -380,5 +382,6 @@ def test_corrections_2d(gauss_data_2d):
-    f    = fitf.fit(fitf.gauss, x, y, (1e5, mean, std))
+    sigma = poisson_sigma(y)
+    f    = fitf.fit(fitf.gauss, x, y, (1e5, mean, std), sigma=sigma)
     assert f.chi2 < 3
```
This `f.chi2 < 3` is maybe a bit soft. I think we can `flaky` these two tests.

gonzaponte reviewed (commented):
Good work! This looks nice!

bpalmeiro committed:
Add get_chi2_and_pvalue function

Gets reduced chi2 and p-value from fit results. Previous one just took
into account the poisson-smeared case, now it is genelarized for every
smearing given by the user. Poisson case is taken by default.

bpalmeiro committed:
Replace fit with fitf for consistency

bpalmeiro committed:
Add tests for get_chi2_and_pvalue

bpalmeiro committed:
Change fit function to meet new chi2 implementation

Be aware, now sigmas are taken as absolute errors, if given, if not,
they remain as relative weights.

bpalmeiro committed:
Add chi2-related specific tests due to changes in fit function

bpalmeiro committed:
Change and harden fit_correction test to meet new implementation

gonzaponte reviewed (approved):
This PR fixes a long-term issue with the definition of the chi2 and the p-value. The calculations have been encapsulated within a function, which is tested. The previous test suite for the rest of the code has also been fixed, improved and extended.

Great job!
